### PR TITLE
Improve search method binding priority

### PR DESCRIPTION
**Pull request by jamesagnew**

### CODE REVIEW

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/SearchMethodBinding.java [hunk 1]
```diff
@@ -129,114 +141,215 @@ public ReturnTypeEnum getReturnType() {
...
+						break;
+					}
 				}
 			}
 		}
-		return true;
+
+		return retVal;
+
+		// FIXME: remove
```
Comment (fil512) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/SearchMethodBinding.java:267]: remove FIXME :-)

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/SearchMethodBinding.java [hunk 2]
```diff
@@ -264,6 +377,8 @@ protected boolean isAddContentLocationHeader() {
 		return false;
 	}
 
+
+	// FIXME: remove
```
Comment (fil512) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/SearchMethodBinding.java:381]: remove FIXME :-)

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/SearchParameter.java [hunk 3]
```diff
@@ -248,42 +249,52 @@ public void setRequired(boolean required) {
...
+					myParamBinder = new FhirPrimitiveBinder((Class<IPrimitiveType<?>>) theType);
 					myParamType = RestSearchParameterTypeEnum.STRING;
 				}
 			}
 		} else {
-			throw new ConfigurationException("Unsupported data type for parameter: " + type.getCanonicalName());
+			throw new ConfigurationException("Unsupported data theType for parameter: " + theType.getCanonicalName());
+		}
+
+		if (DateRangeParam.class.isAssignableFrom(theType)) {
```
Comment (fil512) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/SearchParameter.java:293]: Wouldn't NumberParam and QuantityParam also support repetition?

Comment (fil512) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/SearchParameter.java:293]: Consider changing to:
```
if (theType.supportsRepetition())
```

File: hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/method/MethodMatchEnumTest.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+import static org.junit.Assert.*;
+
+public class MethodMatchEnumTest {
+
+	@Test
+	public void testOrder() {
+		assertEquals(0, MethodMatchEnum.NONE.ordinal());
+		assertEquals(1, MethodMatchEnum.APPROXIMATE.ordinal());
+		assertEquals(2, MethodMatchEnum.PERFECT.ordinal());
```
Comment (fil512) [inline hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/method/MethodMatchEnumTest.java:13]: Is EXACT more descriptive than PERFECT?

File: hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+import static org.junit.Assert.assertEquals;
+
+public class SearchDateParamPriorityTest {
+	private static final Logger ourLog = LoggerFactory.getLogger(SearchDateParamPriorityTest.class);
+
+	private static FhirContext ourCtx = FhirContext.forR4();
+
+	@ClassRule
+	public static RestfulServerRule ourServerRule = new RestfulServerRule(ourCtx);
```
Comment (fil512) [inline hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java:30]: :-) :-)

File: hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+	@Before
+	public void before() {
+		myLastMethod = null;
+	}
+
+	@After
+	public void after() {
+		while (ourServerRule.getRestfulServer().getResourceProviders().size() > 0) {
+			ourServerRule.getRestfulServer().unregisterProvider(ourServerRule.getRestfulServer().getResourceProviders().iterator().next());
```
Comment (fil512) [inline hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java:41]: Can we add an unregisterAllProviders() method to RestfulServer?

File: hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+		client
+			.search()
+			.forResource("Patient")
+			.returnBundle(Bundle.class)
+			.execute();
+
+		assertEquals("find", myLastMethod);
+	}
+
+	public class PatientResourceProvider implements IResourceProvider {
```
Comment (fil512) [inline hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java:93]: This is a really beautiful test.

File: hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	public void after() {
+		while (ourServerRule.getRestfulServer().getResourceProviders().size() > 0) {
+			ourServerRule.getRestfulServer().unregisterProvider(ourServerRule.getRestfulServer().getResourceProviders().iterator().next());
+		}
+	}
+
+	@Test
+	public void testDateRangeSelectedWhenMultipleParametersProvided() {
+		PatientResourceProvider provider = new PatientResourceProvider();
+		ourServerRule.getRestfulServer().registerProviders(provider);
```
Comment (fil512) [inline hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java:48]: Throw the above two lines in a @Before

File: hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java (hunk 5 above)
Comment (jamesagnew) [reply hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java:30]: I am so in love with this approach now.....

File: hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java (hunk 6 above)
Comment (jamesagnew) [reply hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java:41]: Yeah I guess we could really. I'll do that.

File: hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java (hunk 8 above)
Comment (jamesagnew) [reply hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java:48]: Done

File: hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java (hunk 7 above)
Comment (jamesagnew) [reply hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java:93]: Haha thanks

File: hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java (hunk 8 above)
Comment (jamesagnew) [reply hapi-fhir-structures-r4/src/test/java/ca/uhn/fhir/rest/server/SearchDateParamPriorityTest.java:48]: done

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/SearchMethodBinding.java (hunk 2 above)
Comment (jamesagnew) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/SearchMethodBinding.java:381]: done.... CI caught this too, woohoo!

File: hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/method/MethodMatchEnumTest.java (hunk 4 above)
Comment (jamesagnew) [reply hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/method/MethodMatchEnumTest.java:13]: Yeah, that is better.. changing

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/SearchMethodBinding.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/SearchMethodBinding.java]: Done

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/SearchParameter.java (hunk 3 above)
Comment (jamesagnew) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/method/SearchParameter.java]: Good call, DateRangeParam is kind of the outlier, have added tests to show that regular "AndParamXXX" types work too.

**Pull request merged**